web/modules/custom/parser/src/Service/XMLReader.php

```diff
@@ -0,0 +1,226 @@
+use Drupal\node\Entity\Node;
+use Drupal\Component\Plugin\Exception\InvalidPluginDefinitionException;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+use SimpleXMLElement;
+
+/**
+ * Class XMLReader.
+ *
+ * Parse an XML file.
+ */
+class XMLReader {
+
+  protected $entityTypeManager;
+  protected $paragraphStorage;
+
+  /**
+   * XMLReader constructor.
+   *
+   * Needed for the EntityTypeManager dependency injection.
+   *
+   * @param \Drupal\Core\Entity\EntityTypeManagerInterface $entityTypeManager
+   *   The entity type manager interface.
+   *
+   * @throws \Exception
+   */
+  public function __construct(EntityTypeManagerInterface $entityTypeManager) {
+    $this->entityTypeManager = $entityTypeManager;
+    try {
+      $this->paragraphStorage = $this->entityTypeManager->getStorage('paragraph');
+    }
+    catch (InvalidPluginDefinitionException $ipde) {
+      throw new \Exception('Exception on location_telephone paragraph,  ' . $ipde->getMessage());
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   *
+   * @throws \Exception
+   */
+  public static function create(ContainerInterface $container) {
+    return new static(
+      $container->get('entity_type.manager')
+    );
+  }
+
+  /**
+   * Parses xml file with php function SimpleXML.
+   *
+   * @throws \Exception
+   */
+  public function parse($xmlFile) {
+    if (!is_file($xmlFile)) {
+      throw new \RuntimeException(sprintf('File "%s" does not exist.', $xmlFile));
+    }
+    if (!is_readable($xmlFile)) {
+      throw new \RuntimeException(sprintf('File "%s" cannot be read.', $xmlFile));
+    }
+    // Get all telephone paragraphs.
+    $phone_paragraphs = $this->paragraphStorage
+      ->loadByProperties(['type' => 'location_telephone']);
+    $phones = [];
+    foreach ($phone_paragraphs as $phone) {
+      $phones[$phone->id()] = $phone;
+    }
+    // Get all locations entities.
+    try {
+      $locations = $this->entityTypeManager
+        ->getStorage('node')
+        ->loadByProperties(['type' => 'location']);
+    }
+    catch (InvalidPluginDefinitionException $ipde) {
+      throw new \Exception('Exception on location node, ' . $ipde->getMessage());
+    }
+    // Get XML offices to update on Drupal.
+    $xml_offices = simplexml_load_file($xmlFile)->LIST_G_CNSL_ORG_ID->G_CNSL_ORG_ID;
+    // Update each XML offices that is found in Drupal content.
+    foreach ($xml_offices as $xml_office) {
+      $xmlId = (string) $xml_office->CNSL_ORG_ID1;
+      foreach ($locations as $location) {
+        // Search by CNSL org id.
+        $cnslField = $location->get('field_location_cnsl_id')->getValue();
+        if (empty($cnslField)) {
+          continue;
+        }
+        // Loop until the CNSL id is found.
+        if ($xmlId != $cnslField[0]['value']) {
+          continue;
+        }
+        // If the CNSL id is found, update the node and go to the next XML id.
+        try {
+          $this->updateLocationPhones($location, $xml_office, $phones);
```

Comment:
Could this be refactored into one single function that updates all, so you only call entity->save() once?

Reply:
since the entities are different, I cannot do that.